### PR TITLE
feat(vm): Add an option to limit the maximum stack space used by a thread

### LINKED ISSUE #171: Crash in try_gluon
`try_gluon` crashed overnight and unfortunately I made an error when enabling the logging so this is all relevant logging for now. Presumably it has to do with the garbage collector but hopefully it occurs again so I can get some more information.

```
hread '<unnamed>' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/ec2-user/.cargo/registry/src/github.com-1ecc6299db9ec823/gluon_vm-0.2.0/src/thread.rs:1082
note: Run with `RUST_BACKTRACE=1` for a backtrace.
thread '<unnamed>' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/ec2-user/.cargo/registry/src/github.com-1ecc6299db9ec823/gluon_vm-0.2.0/src/thread.rs:1082
thread '<unnamed>' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/ec2-user/.cargo/registry/src/github.com-1ecc6299db9ec823/gluon_vm-0.2.0/src/thread.rs:1082
thread '<unnamed>' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/ec2-user/.cargo/registry/src/github.com-1ecc6299db9ec823/gluon_vm-0.2.0/src/thread.rs:1082
thread '<unnamed>' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/ec2-user/.cargo/registry/src/github.com-1ecc6299db9ec823/gluon_vm-0.2.0/src/thread.rs:1082
thread '<unnamed>' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/ec2-user/.cargo/registry/src/github.com-1ecc6299db9ec823/gluon_vm-0.2.0/src/thread.rs:1082
thread '<unnamed>' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/ec2-user/.cargo/registry/src/github.com-1ecc6299db9ec823/gluon_vm-0.2.0/src/thread.rs:1082
thread '<unnamed>' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/ec2-user/.cargo/registry/src/github.com-1ecc6299db9ec823/gluon_vm-0.2.0/src/thread.rs:1082
error: Process didn't exit successfully: `target/release/main` (signal: 11, SIGSEGV: invalid memory reference)
```

### PR DESCRIPTION
**Pull request by Marwes**
Fixes #171

### CODE REVIEW

File: vm/src/thread.rs [hunk 1]
```diff
@@ -753,10 +761,19 @@ impl<'b> OwnedContext<'b> {
                     let state = {
                         let mut context = context.borrow_mut();
 
+                        let instruction_index = context.stack.frame.instruction_index;
```
Comment (brendanzab) [inline vm/src/thread.rs]: Could assign `closure.function.required_stack_size` here too to reduce the size of the predicate at 768

File: tests/limits.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    let _ = ::env_logger::init();
+    let vm = make_vm();
+    vm.context().set_stack_limit(3);
+    let result = Compiler::new()
+        .run_expr::<OpaqueValue<&Thread, Hole>>(&vm, "example", r#" [1, 2, 3, 4] "#);
+    match result {
+        Err(Error::VM(VMError::StackOverflow(3))) => (),
+        Err(err) => panic!("Unexpected error `{:?}`", err),
+        Ok(_) => panic!("Expected an error"),
+    }
```
Comment (brendanzab) [inline tests/limits.rs]: Spacing this out would be nice:

```rust
    let _ = ::env_logger::init();

    let vm = make_vm();
    vm.context().set_stack_limit(3);

    let compiler = Compiler::new();
    let expr = r#" [1, 2, 3, 4] "#;
    let result = compiler.run_expr::<OpaqueValue<&Thread, Hole>>(&vm, "example", expr);

    match result {
        Err(Error::VM(VMError::StackOverflow(3))) => (),
        Err(err) => panic!("Unexpected error `{:?}`", err),
        Ok(_) => panic!("Expected an error"),
    }
```

File: vm/src/compiler.rs [hunk 3]
```diff
@@ -33,6 +33,8 @@ enum FieldAccess {
 #[derive(Debug)]
 pub struct CompiledFunction {
     pub args: VmIndex,
+    /// The maximum possible number of stack slots needed for this function
+    pub required_stack_size: VmIndex,
```
Comment (brendanzab) [inline vm/src/compiler.rs]: `max_stack_size` or `stack_limit`?

File: vm/src/compiler.rs [hunk 4]
```diff
@@ -118,11 +121,23 @@ impl FunctionEnv {
...
-        debug!("{:?} {} {}", i, self.stack_size, i.adjust());
-        self.stack_size = (self.stack_size as i32 + i.adjust()) as VmIndex;
+        let adjustment = i.adjust();
+        debug!("{:?} {} {}", i, self.stack_size, adjustment);
+        if adjustment > 0 {
+            self.increase_stack(adjustment as VmIndex);
+        } else {
+            self.stack_size -= -adjustment as VmIndex;
+        }
         self.function.instructions.push(i);
```
Comment (brendanzab) [inline vm/src/compiler.rs]: Add some spacing around some of these things to make it less dense. Also change `i` to `instruction`.

File: vm/src/compiler.rs (hunk 3 above)
Comment (Marwes) [inline vm/src/compiler.rs]: `max_stack_size` may be clearer, going to replace uses of `stack_limit` since it makes no indication towards what limit it is.

File: vm/src/thread.rs [hunk 5]
```diff
@@ -241,6 +241,7 @@ impl RootedThread {
                 stack: Stack::new(),
                 record_map: FieldMap::new(),
                 hook: None,
+                stack_size_limit: VmIndex::max_value(),
```
Comment (brendanzab) [inline vm/src/thread.rs]: What is the reasoning for having a different field name here? ie. `limit_` vs `max_`?

File: tests/limits.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+#[test]
+fn stack_overflow() {
+    let _ = ::env_logger::init();
+    let vm = make_vm();
+
+    vm.context().set_stack_size_limit(3);
+
+    let result = Compiler::new()
+        .run_expr::<OpaqueValue<&Thread, Hole>>(&vm, "example", r#" [1, 2, 3, 4] "#);
```
Comment (brendanzab) [inline tests/limits.rs]: I would still prefer these assigned to separate variables, as per https://github.com/Marwes/gluon/pull/173/files#r81462317 - makes the lines a little easier to scan vertically.

File: tests/limits.rs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        Err(Error::VM(VMError::OutOfMemory { limit: 10, .. })) => (),
+        Err(err) => panic!("Unexpected error `{:?}`", err),
+        Ok(_) => panic!("Expected an error"),
+    }
+}
+
+#[test]
+fn stack_overflow() {
+    let _ = ::env_logger::init();
+    let vm = make_vm();
```
Comment (brendanzab) [inline tests/limits.rs]: This is not related to logging, so it should be grouped with `vm.context().set_stack_size_limit(3);`.

File: tests/limits.rs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    vm.set_memory_limit(10);
+    let result = Compiler::new()
+        .run_expr::<OpaqueValue<&Thread, Hole>>(&vm, "example", r#" [1, 2, 3, 4] "#);
+    match result {
+        // FIXME This should just need to match on the explicit out of memory error
+        Err(Error::VM(VMError::OutOfMemory { limit: 10, .. })) => (),
+        Err(err) => panic!("Unexpected error `{:?}`", err),
+        Ok(_) => panic!("Expected an error"),
+    }
```
Comment (brendanzab) [inline tests/limits.rs]: Need to fix up the formatting here to be consistent with the `stack_overflow` test.

File: vm/src/thread.rs (hunk 5 above)
Comment (Marwes) [inline vm/src/thread.rs]: Wanted to distinguish the maximum total allowed stack size from the the maximum required for a specific function, but I may have failed in that respect :/.

File: tests/limits.rs (hunk 8 above)
Comment (Marwes) [inline tests/limits.rs]: Oops, only copied it over.

File: vm/src/thread.rs (hunk 5 above)
Comment (brendanzab) [inline vm/src/thread.rs]: I think it would be implicit for the specific datatype. Doc comments could help though!
